pkgs/development/libraries/glibc/ld-fallback.patch

```diff
@@ -0,0 +1,457 @@
+@@ -2342,6 +2371,9 @@ _dl_rtld_di_serinfo (struct link_map *loader, Dl_serinfo *si, bool counting)
+      Here is where ld.so.cache gets checked, but we don't have
+      a way to indicate that in the results for Dl_serinfo.  */
+ 
++  /* Try the LD_FALLBACK_PATH environment variable.  */
++  add_path (&p, &__rtld_fallback_path_list, XXX_FALLBACK);
++
+   /* Finally, try the default path.  */
+   if (!(loader->l_flags_1 & DF_1_NODEFLIB))
+     add_path (&p, &__rtld_search_dirs, XXX_default);
+diff --git a/elf/dl-main.h b/elf/dl-main.h
+index 92766d06b4..c9d85f4dcc 100644
+--- a/elf/dl-main.h
++++ b/elf/dl-main.h
+@@ -78,6 +78,12 @@ struct dl_main_state
+   /* Where library_path comes from.  LD_LIBRARY_PATH or --library-path.  */
+   const char *library_path_source;
+ 
++  /* The fallback search path.  */
++  const char *fallback_path;
++
++  /* Where fallback_path comes from.  LD_FALLBACK_PATH or --fallback-path.  */
++  const char *fallback_path_source;
++
+   /* The list preloaded objects from LD_PRELOAD.  */
+   const char *preloadlist;
+ 
+@@ -110,6 +116,7 @@ static inline void
+ call_init_paths (const struct dl_main_state *state)
+ {
+   _dl_init_paths (state->library_path, state->library_path_source,
++                  state->fallback_path, state->fallback_path_source,
+                   state->glibc_hwcaps_prepend, state->glibc_hwcaps_mask);
+ }
+ 
+diff --git a/elf/dl-support.c b/elf/dl-support.c
+index 44a54dea07..3b402d53a6 100644
+--- a/elf/dl-support.c
++++ b/elf/dl-support.c
+@@ -288,6 +288,7 @@ _dl_non_dynamic_init (void)
+   /* Initialize the data structures for the search paths for shared
+      objects.  */
+   _dl_init_paths (getenv ("LD_LIBRARY_PATH"), "LD_LIBRARY_PATH",
++		  getenv ("LD_FALLBACK_PATH"), "LD_FALLBACK_PATH",
+ 		  /* No glibc-hwcaps selection support in statically
+ 		     linked binaries.  */
+ 		  NULL, NULL);
+diff --git a/elf/dl-usage.c b/elf/dl-usage.c
+index 98f0b0d027..98b9da57aa 100644
+--- a/elf/dl-usage.c
++++ b/elf/dl-usage.c
+@@ -187,6 +187,8 @@ setting environment variables (which would be inherited by subprocesses).\n\
+   --inhibit-cache       Do not use " LD_SO_CACHE "\n\
+   --library-path PATH   use given PATH instead of content of the environment\n\
+                         variable LD_LIBRARY_PATH\n\
++  --fallback-path PATH  use given PATH instead of content of the environment\n\
++                        variable LD_FALLBACK_PATH\n\
+   --glibc-hwcaps-prepend LIST\n\
+                         search glibc-hwcaps subdirectories in LIST\n\
+   --glibc-hwcaps-mask LIST\n\
+diff --git a/elf/link.h b/elf/link.h
+index 3b5954d981..d114e04095 100644
+--- a/elf/link.h
++++ b/elf/link.h
+@@ -130,6 +130,7 @@ enum
+     LA_SER_LIBPATH = 0x02,	/* Directory from LD_LIBRARY_PATH.  */
+     LA_SER_RUNPATH = 0x04,	/* Directory from RPATH/RUNPATH.  */
+     LA_SER_CONFIG = 0x08,	/* Found through ldconfig.  */
++    LA_SER_FALLBACK = 0x0a,	/* Directory from LD_FALLBACK_PATH.  */
```

Comment:
Any reason to use `0x0a` here? It seems that all other tag values are exact powers of 2. In particular, some random enum just below this one in the original source also satisfies this property:

```c
{
    LA_SYMB_NOPLTENTER = 0x01,        /* la_pltenter will not be called.  */
    LA_SYMB_NOPLTEXIT = 0x02,        /* la_pltexit will not be called.  */
    LA_SYMB_STRUCTCALL = 0x04,        /* Return value is a structure.  */
    LA_SYMB_DLSYM = 0x08,        /* Binding due to dlsym call.  */
    LA_SYMB_ALTVALUE = 0x10        /* Value has been changed by a previous
                                   la_symbind call.  */
  };
  ```

I don't know the glibc enough to know if it relies on this property somewhere, fiddling with bits, but it would feel safer/more uniform to follow this apparent convention here as well:

```suggestion
+    LA_SER_FALLBACK = 0x10,	/* Directory from LD_FALLBACK_PATH.  */
```

Reply:
Fixed.